### PR TITLE
Display webhook calls in contact history

### PR DESCRIPTION
**Pull request by rowanseymour**

### CODE REVIEW

File: temba/api/migrations/0014_auto_20170410_0731.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    operations = [
+        migrations.AddField(
+            model_name='webhookevent',
+            name='run',
+            field=models.ForeignKey(help_text='The flow run that triggered this event', null=True, on_delete=django.db.models.deletion.CASCADE, to='flows.FlowRun'),
+        ),
+        migrations.AlterField(
+            model_name='webhookevent',
+            name='event',
+            field=models.CharField(choices=[('mo_sms', 'Incoming SMS Message'), ('mt_sent', 'Outgoing SMS Sent'), ('mt_dlvd', 'Outgoing SMS Delivered to Recipient'), ('mt_fail', 'Outgoing SMS Failed to be Delivered to Recipient'), ('mt_call', 'Outgoing Call'), ('mt_miss', 'Missed Outgoing Call'), ('mo_call', 'Incoming Call'), ('mo_miss', 'Missed Incoming Call'), ('alarm', 'Channel Alarm'), ('flow', 'Flow Step Reached'), ('categorize', 'Flow Categorization')], help_text='The event type for this event', max_length=16),
```
Comment (rowanseymour) [inline temba/api/migrations/0014_auto_20170410_0731.py]: `mt_fail` was previously not included in this list although a valid value

File: temba/api/migrations/0014_auto_20170410_0731.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+        ),
+        migrations.AlterField(
+            model_name='webhookresult',
+            name='event',
+            field=models.ForeignKey(help_text='The event that this result is tied to',
+                                    on_delete=django.db.models.deletion.CASCADE, related_name='results',
```
Comment (rowanseymour) [inline temba/api/migrations/0014_auto_20170410_0731.py]: Just setting a nicer `related_name` than `webhookresult_set`

File: temba/airtime/models.py [hunk 3]
```diff
@@ -97,7 +96,7 @@ def get_transferto_response(self, **kwargs):
 
     @classmethod
     def trigger_airtime_event(cls, org, ruleset, contact, event):
-
+        from temba.api.models import get_api_user
```
Comment (rowanseymour) [inline temba/airtime/models.py]: Had to rework dependencies so airtime doesn't import from api/flows directly

File: temba/api/models.py [hunk 4]
```diff
@@ -278,48 +280,52 @@ def trigger_flow_event(cls, webhook_url, flow, run, node_uuid, contact, event, a
...
+                if body:
+                    body = body.strip()
+                status_code = response.status_code
             else:
-                response = requests.post(webhook_url, data=data, headers=TEMBA_HEADERS, timeout=10)
-
-            response_text = response.text
-            body = response.text
-            status_code = response.status_code
+                print("!! Skipping WebHook send, SEND_WEBHOOKS set to False")
```
Comment (rowanseymour) [inline temba/api/models.py]: So now in your dev instance, webhooks still won't actually be requested, but we will create results objects and update event objects to look like they were fired.

**Comment by nicpottier** [comment]
Screenshot?

File: temba/api/models.py [hunk 5]
```diff
@@ -278,48 +280,52 @@ def trigger_flow_event(cls, webhook_url, flow, run, node_uuid, contact, event, a
...
+                body = 'Skipped actual send'
+                status_code = 200
 
-            if response.status_code == 200 or response.status_code == 201:
+            if 200 <= status_code < 300:
```
Comment (rowanseymour) [inline temba/api/models.py]: Treat any 2xx response as valid

File: temba/api/models.py [hunk 6]
```diff
@@ -583,9 +576,6 @@ class WebHookResult(SmartModel):
 
     request_time = models.IntegerField(null=True, help_text=_('Time it took to process this request'))
 
-    def stripped_body(self):  # pragma: needs cover
```
Comment (rowanseymour) [inline temba/api/models.py]: Not needed as bodies are always stripped

File: temba/api/views.py [hunk 7]
```diff
@@ -2,37 +2,14 @@
...
 from django.views.generic import View
+from six.moves.urllib.parse import parse_qs
 from smartmin.views import SmartTemplateView, SmartReadView, SmartListView, SmartView
 from temba.channels.models import ChannelEvent
 from temba.orgs.views import OrgPermsMixin
-from urlparse import parse_qs
-from .models import WebHookEvent, WebHookResult, APIToken, Resthook
-
-
-def webhook_status_processor(request):
```
Comment (rowanseymour) [inline temba/api/views.py]: This doesn't seem to be needed - `failed_webhooks`  is only referenced in org_webhook so I added it there, and best I can tell, `webhook_errors_count ` is never used - not even in nyaruka/textit

File: temba/api/models.py [hunk 8]
```diff
@@ -605,21 +595,20 @@ def record_result(cls, event, result):
...
-                                     data=result['data'],
-                                     message=message,
-                                     status_code=result.get('status_code', 503),
-                                     body=result.get('body', None),
-                                     request_time=request_time,
-                                     created_by=api_user,
-                                     modified_by=api_user)
-
-        # keep only the most recent 100 events for each org
-        for old_event in WebHookEvent.objects.filter(org=event.org, status__in=['C', 'F']).order_by('-created_on')[100:]:  # pragma: no cover
```
Comment (rowanseymour) [inline temba/api/models.py]: I'm assuming this should no longer be here now that events are deleted in a task?

**Comment by rowanseymour** [comment]
@nicpottier was this what you were envisaging?

screen shot 2017-04-10 at 14 10 01

**Comment by nicpottier** [comment]
Ya, screen looks good, maybe we should do the same thing as we do for channel logs and show a little paper icon to the right of the timestamp? Also did try testing this with a really long URL to see what it does to the table?

**Comment by nicpottier** [comment]
Eric and I just noticed the bold and font changes on this, Eric prefers the old way, guessing you changed this in the last PR on contact history?

**Comment by rowanseymour** [comment]
@nicpottier seems so but it wasn't intentional - I was confused by this empty summary tag https://github.com/nyaruka/rapidpro/blob/971f47be7b5142faddf9c94c759550edf4ed5f70/templates/contacts/contact_history.haml#L59-L59 and thought the stuff below it was meant to be in it. I'll revert in this PR.

**Comment by rowanseymour** [comment]
@ericnewcomer @nicpottier here's how its looks now - and I've added a long URL. Open to suggestions how best to handle that. Ellipsis?

screen shot 2017-04-11 at 11 28 29

**Comment by nicpottier** [comment]
Ya, pretty ugly with the wrapping, let's do a ... which I think is a CSS tag no?

**Pull request merged**